### PR TITLE
feature: enabling screenshare on Miriway

### PR DESCRIPTION
**Pull request by mattkae**
## How to test
- Build the new snap
- Log in
- Open up firefox
- Screenshare
- Select your output and voila

### CODE REVIEW

File: snap/hooks/install [hunk 1]
```diff
@@ -1,10 +1,16 @@
 #!/bin/sh -e
 
-mkdir -p /usr/share/wayland-sessions/ /usr/lib/systemd/user/ /usr/libexec/
+mkdir -p /usr/share/wayland-sessions/ /usr/lib/systemd/user/ /etc/xdg/xdg-desktop-portal-wlr/ /usr/local/libexec/ /usr/local/share/xdg-desktop-portal/portals/ /usr/local/lib/systemd/user/
```
Comment (AlanGriffiths) [inline snap/hooks/install:3]: Why drop ` /usr/libexec/`? We use it

File: snap/hooks/install [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 sed -e s/miriway-session/miriway.session/ "$SNAP"/usr/share/wayland-sessions/miriway.desktop > /usr/share/wayland-sessions/miriway.desktop
 cp "$SNAP"/usr/lib/systemd/user/miriway-session.target /usr/lib/systemd/user/
 cp "$SNAP"/usr/libexec/miriway-session-startup "$SNAP"/usr/libexec/miriway-session-shutdown /usr/libexec/
 
+# Portal copies
+cp "$SNAP"/etc/xdg/xdg-desktop-portal-wlr/config /etc/xdg/xdg-desktop-portal-wlr/
+cp "$SNAP"/usr/local/libexec/xdg-desktop-portal-wlr /usr/local/libexec/
+cp "$SNAP"/usr/local/share/xdg-desktop-portal/portals/wlr.portal /usr/share/xdg-desktop-portal/portals/
+cp "$SNAP"/usr/local/lib/systemd/user/xdg-desktop-portal-wlr.service /usr/lib/systemd/user/
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply snap/hooks/install:12]: Also, is `/usr/local/...` really appropriate? These are not files built locally on the target system.

File: systemd/usr/libexec/miriway-session-startup [hunk 3]
```diff
@@ -55,6 +55,7 @@
 #  - https://www.freedesktop.org/software/systemd/man/systemd.special.html#graphical-session.target
 #  - https://systemd.io/DESKTOP_ENVIRONMENTS/
 #
+export XDG_CURRENT_DESKTOP=mir:Miriway
```
Comment (AlanGriffiths) [inline systemd/usr/libexec/miriway-session-startup:58]: Why change this? The default configuration has this as "Miriway", if another desktop configuration changed it, should it be overridden?

**Comment by mattkae** [comment]
>     2. As currently proposed it will conflict with unvendored xdg-desktop-portal-wlr on systems that have wlroots based desktops installed. (I note the 24.04 archive has "xdg-desktop-portal-wlr/noble 0.7.1-1build2"). I think some renaming is appropriate to avoid conflicting with other packages installed on the system.
> 
> 
> I would much prefer xdg-desktop-portal-wlr to be packaged elsewhere (ideally, the archive version would "just work"), and if we are maintaining an old version, we should avoid conflicting with the current version.
> 
> Is the alternative a "simple matter" of updating Mir's implementation of `wlr-screencopy` and patching `xdg-desktop-portal-wlr/wlr.portal`?

There are two ways forward if we are to use the archive version. The first is to update Mir to properly use DMABUF and the version of the protocol that relies on it. The other is to fix the xdg-desktop-portal-wlr implementation because it _should_ allow us to not have DMABUFs, but no one really uses it so that it's broken.

Another option would be to create `xdg-desktop-portal-mir` project like most other compositors and just support the portal ourselves over there. At that point, we're basically just implementing the portal though.

Honestly, the best solution is to just implement the portal according to our plan, but that is a lot more work :shrug: 

I was trying to avoid any significant hacking here since this is meant as a "band-aid" solution until we implement the protal

**Comment by AlanGriffiths** [comment]
Yes, I understand this is intended as a "quick and dirty" workaround. But I feel it is a bit too dirty in its current incarnation. At a minimum these need fixing:

1. `/etc/xdg/xdg-desktop-portal-wlr/` is going to conflict with the archive version
2. Dumping stuff in `/usr/local/` is a lie. Moving to system locations is admittedly dubious too (c.f. 1), but with suitable renames would be fine.

I still wonder if its own package (Snap? PPA?) would be more flexible.

@Conan-Kudo you likely have another perspective on packaging this?

**Comment by Conan-Kudo** [comment]
> Another option would be to create `xdg-desktop-portal-mir` project like most other compositors and just support the portal ourselves over there. At that point, we're basically just implementing the portal though.

Actually, I've been working on starting a project based on the discussion at Ubuntu Summit that could be shared across unrelated compositors that use only protocols from w-p (the project will be called `xdg-desktop-portal-wpext`). There are several portals that require compositor integration that efforts to define ext protocols to handle them are in progress. @serebit, @drakulix, and myself have been actively marshaling protocols into place for this.

However...

> Is the alternative a "simple matter" of updating Mir's implementation of `wlr-screencopy`

This is a prerequisite for the new portal implementation. https://github.com/canonical/mir/issues/3659 and https://github.com/canonical/mir/issues/3660 were filed specifically as prep for this new project.

**Comment by AlanGriffiths** [comment]
OK, I don't currently propose to land this: It isn't strategic and it is messy. At most, this PR should be tidied enough to avoid the installation conflicting with `xdg-desktop-portal-wlr`.

Folks that can't wait for the strategic solution can install the `edge/pr137` branch of the Miriway snap.

**Comment by mattkae** [comment]
> OK, I don't currently propose to land this: It isn't strategic and it is messy. At most, this PR should be tidied enough to avoid the installation conflicting with `xdg-desktop-portal-wlr`.
> 
> Folks that can't wait for the strategic solution can install the `edge/pr137` branch of the Miriway snap.

Yeah I agree that this is ill-conceived. It makes sense to keep this on a "bandaid" branch and only point users who really really want it over here.

We will take stronger action when we actually implement portals ;)

**Comment by Conan-Kudo** [comment]
@mattkae this probably makes sense as a separate repo in the miriway github organization.

**Comment by mattkae** [comment]
> @mattkae this probably makes sense as a separate repo (and not necessarily in the miriway github organization).

This still makes sense here as a temporary band-aid for those who need to use it, but I don't think it should be landed.

If anything, this PR helps us to understand the problem-space from Mir's perspective a bit more and can hopefully inform a good implementation in the future.

File: systemd/usr/libexec/miriway-session-startup [hunk 4]
```diff
@@ -57,7 +57,8 @@
 #
 export XDG_SESSION_DESKTOP="${XDG_SESSION_DESKTOP:-Miriway}"
 export XDG_SESSION_TYPE=wayland
-VARIABLES="DESKTOP_SESSION XDG_CURRENT_DESKTOP XDG_SESSION_DESKTOP XDG_SESSION_TYPE DISPLAY WAYLAND_DISPLAY"
+export XDG_DESKTOP_PORTAL_DIR="$SNAP/usr/share/xdg-desktop-portal/portals/"
```
Comment (AlanGriffiths) [inline systemd/usr/libexec/miriway-session-startup:60]: As mentioned by @Conan-Kudo on Matrix, this script isn't guaranteed to execute in a Snap context.

It is probably best to create a `snap-utils` wrapper script that sets `XDG_DESKTOP_PORTAL_DIR`, runs this startup script, then runs the portal binaries

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]

**Comment by Conan-Kudo** [comment]
If we land this, we probably want to install the portal configuration files somewhere in our own miriway configuration directory paths to not collide with the actual wlr portal files. In addition to to that, this needs to be a cmake install configuration knob, like the SDDM one.

**Comment by AlanGriffiths** [comment]
> If we land this, we probably want to install the portal configuration files somewhere in our own miriway configuration directory paths to not collide with the actual wlr portal files

The way this PR is now going is "only in the snap recipe". The snap already includes more than the base install. (And I know need to provide some more documentation around this difference.)

**Comment by Conan-Kudo** [comment]
Then it needs to be guarded with a snap cmake configuration knob.

File: systemd/usr/libexec/miriway-session-startup (hunk 4 above)
Comment (Saviq) [reply systemd/usr/libexec/miriway-session-startup:60]: The environment can probably just go into `snapcraft.yaml`?

File: systemd/usr/share/xdg-desktop-portal/portals/wlr.portal [hunk 5]
```diff
@@ -0,0 +1,4 @@
+[portal]
+DBusName=org.freedesktop.impl.portal.desktop.wlr
+Interfaces=org.freedesktop.impl.portal.Screenshot;org.freedesktop.impl.portal.ScreenCast;
+UseIn=wlroots;sway;Wayfire;river;phosh;Hyprland;Miriway
```
Comment (Saviq) [inline systemd/usr/share/xdg-desktop-portal/portals/wlr.portal:4]:
```suggestion
UseIn=Mir
```

Comment (Saviq) [reply systemd/usr/share/xdg-desktop-portal/portals/wlr.portal:4]: And then make `XDG_CURRENT_DESKTOP=Miriway:Mir`, this way snap downstreams can benefit from this straight up?

File: systemd/usr/libexec/miriway-session-startup [hunk 6]
```diff
@@ -115,3 +116,7 @@ systemctl --user reset-failed
 # shellcheck disable=SC2086
 systemctl --user import-environment $VARIABLES
 systemctl --user start "$SESSION_TARGET"
+
+# Run portals
+systemd-run --user $SNAP/usr/libexec/xdg-desktop-portal --replace
+systemd-run --user $SNAP/usr/local/libexec/xdg-desktop-portal-wlr --replace -c $SNAP/etc/xdg/xdg-desktop-portal-wlr/config
```
Comment (Saviq) [inline systemd/usr/libexec/miriway-session-startup:122]: Yeah these should only run in the snapped session.

**Review by Saviq — CHANGES_REQUESTED** [review]

File: systemd/usr/share/xdg-desktop-portal/portals/wlr.portal (hunk 5 above)
Comment (Saviq) [reply systemd/usr/share/xdg-desktop-portal/portals/wlr.portal:4]: Miracle should set `XDG_CURRENT_DESKTOP=Miracle:Mir`, no? It's not matched by `startswith`, AIUI?

Or do you mean it already has `:mir`?

Comment (Saviq) [reply systemd/usr/share/xdg-desktop-portal/portals/wlr.portal:4]: I suppose we could just rename it `mir.portal` and install outside of snaps, too

File: systemd/usr/libexec/miriway-session-startup.in [hunk 7]
```diff
@@ -115,3 +116,9 @@ systemctl --user reset-failed
 # shellcheck disable=SC2086
 systemctl --user import-environment $VARIABLES
 systemctl --user start "$SESSION_TARGET"
+
+if [ "$IS_SNAP_BUILD" = "ON" ]; then
+    # Run portals
+    systemd-run --user $SNAP/usr/libexec/xdg-desktop-portal --replace
+    systemd-run --user $SNAP/usr/local/libexec/xdg-desktop-portal-wlr --replace -c $SNAP/etc/xdg/xdg-desktop-portal-wlr/config
+fi
```
[the remark this answers is not in the document]
Comment (Saviq) [reply systemd/usr/libexec/miriway-session-startup.in:124]: Like, you can `organize:` this one to one with an `.orig` suffix, and `source` it in your snap-custom one.

File: systemd/usr/share/xdg-desktop-portal/portals/mir.portal [hunk 8]
```diff
@@ -0,0 +1,4 @@
+[portal]
+DBusName=org.freedesktop.impl.portal.desktop.wlr
+Interfaces=org.freedesktop.impl.portal.Screenshot;org.freedesktop.impl.portal.ScreenCast;
+UseIn=mir
```
[the remark this answers is not in the document]
Comment (mattkae) [reply systemd/usr/share/xdg-desktop-portal/portals/mir.portal:4]: To be honest, `UseIn` is deprecated so it technically does not matter :) But I will do!

**Comment by AlanGriffiths** [comment]
> There is some "pre-existing" about that because both cmake and the snap install hook put `miriway-session-startup` on the host system. But at present they are at least identical.

Proposed a fixed for my mess: #139

File: systemd/usr/libexec/miriway-session-startup-snap
Comment (AlanGriffiths) [inline systemd/usr/libexec/miriway-session-startup-snap:1]: I would prefer this in `snap-utils`

File: systemd/usr/libexec/miriway-session-startup-snap [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh -eu
+
+bindir=$(dirname "$0")
+if [ "${bindir}" != "" ]; then bindir="${bindir}/"; fi
+
+"${bindir}"miriway-session-startup "$@"
```
Comment (AlanGriffiths) [inline systemd/usr/libexec/miriway-session-startup-snap:6]:
```suggestion
source "$SNAP/usr/libexec/miriway-session-startup" "$@"
```

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Nearly there

File: systemd/usr/libexec/miriway-session-startup-snap (hunk 9 above)
Comment (mattkae) [reply systemd/usr/libexec/miriway-session-startup-snap]: Fixed! Using `source` or `exec` made it not work though

File: snap-utils/usr/libexec/miriway-session-startup-snap [hunk 10]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh -eu
+
+source "${SNAP}"/usr/libexec/miriway-session-startup "$@"
```
Comment (AlanGriffiths) [inline snap-utils/usr/libexec/miriway-session-startup-snap:3]: @mattkae you commented that "Using source or exec made it not work though". So is this good now?

**Review by AlanGriffiths — APPROVED** [review]
Looks good to me.

One point to clarify, and maybe @Saviq wants to chip in again?

File: snap-utils/usr/libexec/miriway-session-startup-snap (hunk 10 above)
Comment (Saviq) [reply snap-utils/usr/libexec/miriway-session-startup-snap]: Confused. Sourcing or executing shell script from another shell script doesn't work, but forking it, does…?

Sounds like systemd expects a fork somewhere?

Comment (AlanGriffiths) [reply snap-utils/usr/libexec/miriway-session-startup-snap]: The reason is "source: not found".

But `.` works

File: snap-utils/usr/libexec/miriway-session-startup-snap [hunk 11]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh -eu
+
+"${SNAP}"/usr/libexec/miriway-session-startup "$@"
```
Comment (AlanGriffiths) [inline snap-utils/usr/libexec/miriway-session-startup-snap:3]:
```suggestion
. "${SNAP}"/usr/libexec/miriway-session-startup "$@"
```

File: snap-utils/usr/libexec/miriway-session-startup-snap (hunk 10 above)
Comment (Saviq) [reply snap-utils/usr/libexec/miriway-session-startup-snap]: I was tricksed by `exec` not working :P

**Comment by Conan-Kudo** [comment]
Shouldn't the portal configuration files be in the snap subdirectory? It's not useful outside of that.

**Comment by Saviq** [comment]
> It's not useful outside of that.

Why not?

**Comment by Conan-Kudo** [comment]
The files have hardcoded snap paths in them.

**Comment by Saviq** [comment]
> The files have hardcoded snap paths in them.

This looks good to me? Agree the `config` file should go to `snap-utils`, though.

https://github.com/Miriway/Miriway/blob/8008944c815e12b46d2dea220433ddd7cd076f5a/systemd/usr/share/xdg-desktop-portal/portals/mir.portal#L1-L4

**Comment by Conan-Kudo** [comment]
That one is fine, yeah.

**Comment by tarek-y-ismail** [comment]
If `slurp` reports that the compositor doesn't support `zwlr_layer_shell_v1`, make sure to add `add-wayland-extensions=zwlr_layer_shell_v1` and NOT `add-wayland-extension=...` (singular "extension")  to your miriway config. I was confused for a bit since I had added `shell-add-wayland-extension=...` (singular "extension") earlier while debugging and assumed the non-shell option would follow the same pattern.

**Pull request merged**